File: patterns/streaming/openai-like/javascript/stream-chat.ts
import 'dotenv/config';
import OpenAI from 'openai';

/**
 * Demonstrates streaming chat completions with OpenAI.
 */
async function main() {
    const client = new OpenAI({
        apiKey: process.env.OPENAI_API_KEY,
    });

    console.log('Sending streaming request to OpenAI...\n');
    console.log('--- Stream Start ---');

    try {
        const stream = await client.chat.completions.create({
            model: 'gpt-3.5-turbo',
            messages: [{ role: 'user', content: 'Write a short poem about the ocean.' }],
            stream: true,
        });

        for await (const chunk of stream) {
            const content = chunk.choices[0]?.delta?.content || '';
            process.stdout.write(content);
        }

        console.log('\n--- Stream End ---');
    } catch (error) {
        console.error('Error:', error);
    }
}

main();

File: patterns/streaming/openai-like/python/stream_chat.py
import os
import sys
from openai import OpenAI
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

def main():
    """
    Demonstrates streaming chat completions with OpenAI.
    """
    api_key = os.environ.get("OPENAI_API_KEY")
    client = OpenAI(api_key=api_key)

    print("Sending streaming request to OpenAI...\n")

    stream = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "user", "content": "Write a short poem about the ocean."}
        ],
        stream=True,
    )

    print("--- Stream Start ---")
    for chunk in stream:
        content = chunk.choices[0].delta.content
        if content:
            sys.stdout.write(content)
            sys.stdout.flush()
    print("\n--- Stream End ---")

if __name__ == "__main__":
    main()
